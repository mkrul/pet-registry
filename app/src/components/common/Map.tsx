import React, { useRef, useState, useEffect } from "react";
import { MapContainer, TileLayer, useMapEvents, Marker } from "react-leaflet";
import L from "leaflet";
import "leaflet/dist/leaflet.css";
import Spinner from "./Spinner";
import Notification from "./Notification";
import { MapProps } from "../../types/common/Map";
import { isUSLocation } from "../../utils/locationUtils";
import { NotificationType } from "../../types/common/Notification";
import "../../utils/leafletSetup";
import { MapLocation } from "../../types/common/Map";
import { LocationData } from "../../types/Report";
import { LatLngExpression } from "leaflet";

const findNearestArea = async (
  lat: number,
  lng: number,
  onNotification: (notification: { type: NotificationType; message: string }) => void
): Promise<string> => {
  try {
    const response = await fetch(
      `https://nominatim.openstreetmap.org/reverse?` +
        `format=json&` +
        `lat=${lat}&` +
        `lon=${lng}&` +
        `zoom=10&` +
        `addressdetails=1`
    );

    const data = await response.json();

    if (data && data.address) {
      const areaName =
        data.address.city ||
        data.address.area ||
        data.address.town ||
        data.address.village ||
        data.address.suburb ||
        data.address.neighbourhood ||
        data.address.municipality ||
        data.address.hamlet ||
        data.address.isolated_dwelling ||
        data.address.county;

      if (areaName && !areaName.toLowerCase().includes("unknown")) {
        return areaName;
      }

      const widerResponse = await fetch(
        `https://nominatim.openstreetmap.org/reverse?` +
          `format=json&` +
          `lat=${lat}&` +
          `lon=${lng}&` +
          `zoom=8&` +
          `addressdetails=1`
      );

      const widerData = await widerResponse.json();

      if (widerData && widerData.address) {
        const widerAreaName =
          widerData.address.city ||
          widerData.address.area ||
          widerData.address.town ||
          widerData.address.village ||
          widerData.address.suburb ||
          widerData.address.neighbourhood ||
          widerData.address.municipality ||
          widerData.address.hamlet ||
          widerData.address.isolated_dwelling ||
          widerData.address.county;

        if (widerAreaName && !widerAreaName.toLowerCase().includes("unknown")) {
          return widerAreaName;
        }
      }
    }
    return "Unknown Location";
  } catch (error) {
    onNotification({
      type: NotificationType.ERROR,
      message:
        "The map does not recognize the location you selected. Please choose a different location."
    });
    return "Unknown Location";
  }
};

const findNearbyStreets = async (lat: number, lng: number): Promise<string | null> => {
  try {
    const mainResponse = await fetch(
      `https://nominatim.openstreetmap.org/reverse?` +
        `format=json&` +
        `lat=${lat}&` +
        `lon=${lng}&` +
        `zoom=18&` +
        `addressdetails=1`
    );

    const mainData = await mainResponse.json();
    const mainRoad = mainData?.address?.road;

    if (!mainRoad) {
      return null;
    }

    const radius = 100;
    const query = `
      [out:json][timeout:25];
      (
        way(around:${radius},${lat},${lng})[highway~"^(primary|secondary|tertiary|residential|unclassified)$"][name];
      );
      out body;
      >;
      out skel qt;
    `;

    const overpassResponse = await fetch(
      `https://overpass-api.de/api/interpreter?data=${encodeURIComponent(query)}`
    );

    const overpassData = await overpassResponse.json();

    const roads = overpassData.elements
      .filter(
        (element: { tags?: { name?: string; highway?: string } }) =>
          element.tags?.name && element.tags.name.toLowerCase() !== mainRoad.toLowerCase()
      )
      .map((element: { tags?: { name?: string } }) => element.tags?.name)
      .filter((name?: string): name is string => !!name);

    if (roads.length === 0) {
      const widerQuery = query.replace(`${radius}`, "200");
      const widerResponse = await fetch(
        `https://overpass-api.de/api/interpreter?data=${encodeURIComponent(widerQuery)}`
      );
      const widerData = await widerResponse.json();

      roads.push(
        ...widerData.elements
          .filter(
            (element: { tags?: { name?: string; highway?: string } }) =>
              element.tags?.name && element.tags.name.toLowerCase() !== mainRoad.toLowerCase()
          )
          .map((element: { tags?: { name?: string } }) => element.tags?.name)
          .filter((name?: string): name is string => !!name)
      );
    }

    if (roads.length > 0) {
      return `${mainRoad} at ${roads[0]}`;
    }

    return null;
  } catch (error) {
    console.error("Error finding nearby streets:", error);
    return null;
  }
};

interface MapEventsProps extends MapProps {
  onNotification: (notification: { type: NotificationType; message: string } | null) => void;
}

const MapEvents: React.FC<MapEventsProps> = ({
  onLocationSelect,
  initialLocation,
  onNotification,
  readOnly,
  initialZoom
}) => {
  const [isProcessing, setIsProcessing] = useState(false);
  const [selectedPosition, setSelectedPosition] = useState<[number, number] | null>(
    initialLocation?.latitude && initialLocation?.longitude
      ? [initialLocation.latitude, initialLocation.longitude]
      : null
  );
  const [hasSetInitialView, setHasSetInitialView] = useState(false);

  const map = useMapEvents({
    click: async e => {
      if (!readOnly) {
        setSelectedPosition([e.latlng.lat, e.latlng.lng]);
        await handleLocationSelect(e.latlng.lat, e.latlng.lng);
      }
    }
  });

  const handleLocationSelect = async (lat: number, lng: number, skipLocationFetch = false) => {
    if (isProcessing) return;
    setIsProcessing(true);

    try {
      const formattedLat = Number(lat.toFixed(6));
      const formattedLng = Number(lng.toFixed(6));

      if (skipLocationFetch && initialLocation) {
        const locationData: LocationData = {
          latitude: formattedLat,
          longitude: formattedLng,
          area: initialLocation.area ?? "",
          state: initialLocation.state ?? "",
          country: initialLocation.country ?? "",
          intersection: initialLocation.intersection ?? ""
        };
        onLocationSelect(locationData);
        return;
      }

      const response = await fetch(
        `https://nominatim.openstreetmap.org/reverse?format=json&lat=${formattedLat}&lon=${formattedLng}&zoom=18&addressdetails=1`
      );
      const data = await response.json();
      const address = data.address;

      if (!isUSLocation(address.country || "")) {
        onNotification({
          type: NotificationType.ERROR,
          message: "Sorry, we are only able to support US locations at this time."
        });
        return;
      }

      onNotification(null);

      let area =
        address.area ||
        address.town ||
        address.village ||
        address.suburb ||
        address.municipality ||
        address.neighbourhood;

      if (!area || area === "Unknown") {
        area = await findNearestArea(formattedLat, formattedLng, onNotification);
      }
      const intersectionStr = await findNearbyStreets(formattedLat, formattedLng);

      const locationData: LocationData = {
        latitude: formattedLat,
        longitude: formattedLng,
        area: area ?? "",
        state: address.state ?? "",
        country: address.country ?? "",
        intersection: intersectionStr ?? ""
      };

      if (onLocationSelect) {
        onLocationSelect(locationData);
      }
    } catch (error) {
      console.error("Error handling location:", error);
    } finally {
      setIsProcessing(false);
    }
  };

  useEffect(() => {
    if (initialLocation?.latitude && initialLocation?.longitude) {
      setSelectedPosition([initialLocation.latitude, initialLocation.longitude]);
      map.setView([initialLocation.latitude, initialLocation.longitude], 16, {
        animate: true
      });
      if (!hasSetInitialView) {
        setHasSetInitialView(true);
      }
    } else if (!hasSetInitialView) {
      map.setView([39.8283, -98.5795], initialZoom, { animate: true });
      setHasSetInitialView(true);
    }
  }, [map, initialLocation?.latitude, initialLocation?.longitude, initialZoom, hasSetInitialView]);

  return (
    <>
      {isProcessing ? (
        <div className="absolute inset-0 bg-white/75 z-[1000] flex items-center justify-center">
          <Spinner size={32} bgFaded={false} inline={true} className="text-gray-300" />
        </div>
      ) : null}
      {!readOnly && selectedPosition && <Marker position={selectedPosition} />}
    </>
  );
};

export const Map: React.FC<MapProps> = ({
  onLocationSelect,
  initialLocation,
  initialZoom = 4,
  readOnly = false
}) => {
  const [isLoading, setIsLoading] = useState(true);
  const [notification, setNotification] = useState<{
    type: NotificationType;
    message: string;
  } | null>(null);

  useEffect(() => {
    setIsLoading(false);
  }, []);

  const handleMapClick = (e: L.LeafletMouseEvent) => {
    const location: LocationData = {
      latitude: e.latlng.lat,
      longitude: e.latlng.lng,
      area: "",
      state: "",
      country: "",
      intersection: ""
    };
    onLocationSelect(location);
  };

  return (
    <div className="space-y-2">
      {notification && (
        <Notification
          type={notification.type}
          message={notification.message}
          onClose={() => setNotification(null)}
        />
      )}
      <div className="relative w-full h-[400px] rounded-lg overflow-hidden">
        <MapContainer
          center={
            initialLocation?.latitude && initialLocation?.longitude
              ? ([initialLocation.latitude, initialLocation.longitude] as LatLngExpression)
              : ([39.8283, -98.5795] as LatLngExpression)
          }
          zoom={initialZoom}
          className="h-full w-full"
        >
          <TileLayer
            url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
            attribution='© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
          />
          <MapEvents
            onLocationSelect={onLocationSelect}
            initialLocation={initialLocation}
            onNotification={setNotification}
            readOnly={readOnly}
            initialZoom={initialZoom}
          />
        </MapContainer>
        {isLoading && (
          <div className="absolute inset-0 bg-white/75 z-[1000] flex items-center justify-center">
            <Spinner size={32} bgFaded={false} inline={true} className="text-gray-300" />
          </div>
        )}
      </div>
    </div>
  );
};

export default Map;
