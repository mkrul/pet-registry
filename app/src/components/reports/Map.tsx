import React from "react";
import { MapContainer, TileLayer, Marker, useMapEvents } from "react-leaflet";
import "leaflet/dist/leaflet.css";
import L from "leaflet";

interface MapProps {
  onLocationSelect: (location: {
    latitude: number;
    longitude: number;
    city: string;
    state: string;
    country: string;
  }) => void;
}

const LocationMarker: React.FC<{ onLocationSelect: MapProps["onLocationSelect"] }> = ({
  onLocationSelect
}) => {
  const [position, setPosition] = React.useState<L.LatLng | null>(null);

  useMapEvents({
    click(e) {
      setPosition(e.latlng);
      fetchLocationDetails(e.latlng.lat, e.latlng.lng, onLocationSelect);
    },
    dragend() {
      if (position) {
        fetchLocationDetails(position.lat, position.lng, onLocationSelect);
      }
    }
  });

  return position === null ? null : <Marker position={position} draggable={true} />;
};

const fetchLocationDetails = async (
  lat: number,
  lng: number,
  onLocationSelect: MapProps["onLocationSelect"]
) => {
  try {
    const response = await fetch(
      `https://nominatim.openstreetmap.org/reverse?format=json&lat=${lat}&lon=${lng}&zoom=18&addressdetails=1`
    );
    const data = await response.json();
    const address = data.address;
    onLocationSelect({
      latitude: lat,
      longitude: lng,
      city: address.city || address.town || address.village || "",
      state: address.state || "",
      country: address.country || ""
    });
  } catch (error) {
    console.error("Error fetching location details:", error);
  }
};

const Map: React.FC<MapProps> = ({ onLocationSelect }) => {
  return (
    <div className="w-full h-[400px] rounded-lg overflow-hidden border border-gray-300">
      <MapContainer center={[37.0902, -95.7129]} zoom={4} style={{ width: "100%", height: "100%" }}>
        <TileLayer
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
          attribution='Â© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
        />
        <LocationMarker onLocationSelect={onLocationSelect} />
      </MapContainer>
    </div>
  );
};

export default Map;
