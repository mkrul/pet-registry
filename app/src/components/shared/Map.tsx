import React, { useRef } from "react";
import { MapContainer, TileLayer, Marker, useMapEvents } from "react-leaflet";
import L from "leaflet";
import "leaflet/dist/leaflet.css";

interface MapProps {
  onLocationSelect: (location: {
    latitude: number;
    longitude: number;
    city: string;
    state: string;
    country: string;
  }) => void;
}

const MapEvents = ({ onLocationSelect }: MapProps) => {
  const markerRef = useRef<L.Marker | null>(null);

  const map = useMapEvents({
    click: async e => {
      const { lat, lng } = e.latlng;

      // Remove existing marker if any
      if (markerRef.current) {
        markerRef.current.remove();
      }

      // Create new marker
      markerRef.current = L.marker([lat, lng], { draggable: true }).addTo(map);

      // Get address details from coordinates using Nominatim
      try {
        console.log("Fetching location details for", { lat, lng });
        const response = await fetch(
          `https://nominatim.openstreetmap.org/reverse?format=json&lat=${lat}&lon=${lng}&zoom=18&addressdetails=1`
        );
        const data = await response.json();
        console.log("Received location data:", data);

        const address = data.address;
        onLocationSelect({
          latitude: lat,
          longitude: lng,
          city: address.city || address.town || address.village || "",
          state: address.state || "",
          country: address.country || ""
        });
      } catch (error) {
        console.error("Error fetching location details:", error);
      }

      // Add dragend event listener to marker
      markerRef.current.on("dragend", async () => {
        const position = markerRef.current!.getLatLng();
        try {
          console.log("Fetching location details after drag for", position);
          const response = await fetch(
            `https://nominatim.openstreetmap.org/reverse?format=json&lat=${position.lat}&lon=${position.lng}&zoom=18&addressdetails=1`
          );
          const data = await response.json();
          console.log("Received location data after drag:", data);

          const address = data.address;
          onLocationSelect({
            latitude: position.lat,
            longitude: position.lng,
            city: address.city || address.town || address.village || "",
            state: address.state || "",
            country: address.country || ""
          });
        } catch (error) {
          console.error("Error fetching location details:", error);
        }
      });
    }
  });

  return null;
};

const Map: React.FC<MapProps> = ({ onLocationSelect }) => {
  return (
    <MapContainer center={[37.0902, -95.7129]} zoom={4} style={{ width: "100%", height: "100%" }}>
      <TileLayer
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
        attribution='Â© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
      />
      <MapEvents onLocationSelect={onLocationSelect} />
    </MapContainer>
  );
};

export default Map;
