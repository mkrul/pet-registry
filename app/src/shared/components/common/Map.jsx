import React from "react";
import { MapContainer as LeafletMapContainer, TileLayer } from "react-leaflet";
import "leaflet/dist/leaflet.css";
import "../../utils/leafletSetup";
import { MapEvents } from "./MapEvents";
import { MAP_ZOOM_LEVELS } from "../../constants/map";

export const Map = ({
  onLocationSelect,
  initialLocation,
  readOnly,
  initialZoom,
  onProcessingStateChange,
  showInitialMarker = true
}) => {
  return (
    <div className="h-[400px] w-full rounded-lg overflow-hidden">
      <LeafletMapContainer
        center={[39.8283, -98.5795]}
        zoom={initialZoom || MAP_ZOOM_LEVELS.DEFAULT}
        className="h-full"
      >
        <TileLayer
          attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
        />
        <MapEvents
          onLocationSelect={onLocationSelect}
          initialLocation={initialLocation}
          readOnly={readOnly}
          initialZoom={initialZoom}
          onProcessingStateChange={onProcessingStateChange}
          showInitialMarker={showInitialMarker}
        />
      </LeafletMapContainer>
    </div>
  );
};

export default Map;
