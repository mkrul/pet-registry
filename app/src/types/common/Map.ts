export interface MapProps {
  onLocationSelect?: (location: {
    latitude: number;
    longitude: number;
    area: string;
    state: string;
    country: string;
  }) => void;
  initialLocation?: {
    latitude: number | null;
    longitude: number | null;
  };
  initialZoom?: number;
  readOnly?: boolean;
}

export interface MapLocation {
  latitude: number;
  longitude: number;
  area?: string;
  state?: string;
  country?: string;
}
