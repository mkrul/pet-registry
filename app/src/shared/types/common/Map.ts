import { LocationData } from "../redux/features/reports/ReportsApi";

export interface MapCoordinates {
  latitude: number;
  longitude: number;
}

export interface MapProps {
  onLocationSelect: (location: LocationData) => void;
  initialLocation?: Partial<LocationData & MapCoordinates>;
  initialZoom: number;
  readOnly?: boolean;
}
