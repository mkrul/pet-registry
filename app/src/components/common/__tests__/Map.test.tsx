import { describe, it, expect, vi, beforeEach, type MockInstance } from "vitest";
import { render, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { useMapEvents } from "react-leaflet";
import Map from "../Map";

// Mock react-leaflet components
vi.mock("react-leaflet", () => {
  const actual = vi.importActual("react-leaflet");
  return {
    ...actual,
    MapContainer: ({ children, center, zoom }: any) => (
      <div data-testid="map-container" data-center={JSON.stringify(center)} data-zoom={zoom}>
        {children}
      </div>
    ),
    TileLayer: ({ url, attribution }: any) => (
      <div data-testid="tile-layer" data-url={url} data-attribution={attribution} />
    ),
    Marker: ({ position }: any) => (
      <div data-testid="map-marker" data-position={JSON.stringify(position)} />
    ),
    useMapEvents: vi.fn()
  };
});

// Mock leaflet
vi.mock("leaflet", () => ({
  default: {
    marker: () => ({
      addTo: vi.fn(),
      remove: vi.fn()
    })
  }
}));

// Mock Spinner component
vi.mock("../Spinner", () => ({
  default: ({ size, className }: any) => (
    <div data-testid="spinner" data-size={size} className={className} />
  )
}));

describe("Map", () => {
  const mockOnLocationSelect = vi.fn();
  const mockLocation = {
    latitude: 40.7128,
    longitude: -74.006,
    area: "New York",
    state: "NY",
    country: "USA"
  };

  beforeEach(() => {
    vi.clearAllMocks();
    global.fetch = vi.fn();
  });

  it("renders map with default center when no initial location", () => {
    render(<Map onLocationSelect={mockOnLocationSelect} />);

    const mapContainer = screen.getByTestId("map-container");
    expect(mapContainer).toBeDefined();
    expect(mapContainer.dataset.center).toBe(JSON.stringify([39.8283, -98.5795]));
    expect(mapContainer.dataset.zoom).toBe("15");
  });

  it("renders map with initial location when provided", () => {
    render(<Map onLocationSelect={mockOnLocationSelect} initialLocation={mockLocation} />);

    const mapContainer = screen.getByTestId("map-container");
    expect(mapContainer).toBeDefined();
    expect(mapContainer.dataset.center).toBe(
      JSON.stringify([mockLocation.latitude, mockLocation.longitude])
    );
  });

  it("shows marker when initial location is provided", () => {
    render(<Map onLocationSelect={mockOnLocationSelect} initialLocation={mockLocation} />);

    const marker = screen.getByTestId("map-marker");
    expect(marker).toBeDefined();
    expect(marker.dataset.position).toBe(
      JSON.stringify([mockLocation.latitude, mockLocation.longitude])
    );
  });

  it("shows loading spinner initially", () => {
    render(<Map onLocationSelect={mockOnLocationSelect} />);

    const spinner = screen.getByTestId("spinner");
    expect(spinner).toBeDefined();
    expect(spinner).toHaveClass("text-gray-300");
  });

  it("renders in read-only mode correctly", () => {
    render(
      <Map onLocationSelect={mockOnLocationSelect} initialLocation={mockLocation} readOnly={true} />
    );

    const mapEvents = screen.queryByTestId("map-events");
    expect(mapEvents).toBeNull();
  });

  it("handles location selection correctly", async () => {
    const mockGeocodingResponse = {
      address: {
        area: "Manhattan",
        state: "New York",
        country: "USA"
      }
    };

    (global.fetch as unknown as MockInstance).mockResolvedValueOnce({
      json: () => Promise.resolve(mockGeocodingResponse)
    });

    let clickHandler: (e: any) => Promise<void>;
    (useMapEvents as unknown as MockInstance).mockImplementation(({ click }) => {
      clickHandler = click;
      return { remove: vi.fn() };
    });

    render(<Map onLocationSelect={mockOnLocationSelect} />);

    // Simulate map click
    await act(async () => {
      await clickHandler({ latlng: { lat: 40.7128, lng: -74.006 } });
    });

    expect(global.fetch).toHaveBeenCalledWith(
      expect.stringContaining("nominatim.openstreetmap.org")
    );
    expect(mockOnLocationSelect).toHaveBeenCalledWith({
      latitude: 40.7128,
      longitude: -74.006,
      area: "Manhattan",
      state: "New York",
      country: "USA"
    });
  });

  it("handles geocoding errors gracefully", async () => {
    (global.fetch as unknown as MockInstance).mockRejectedValueOnce(new Error("Geocoding failed"));

    let clickHandler: (e: any) => Promise<void>;
    (useMapEvents as unknown as MockInstance).mockImplementation(({ click }) => {
      clickHandler = click;
      return { remove: vi.fn() };
    });

    render(<Map onLocationSelect={mockOnLocationSelect} />);

    await act(async () => {
      await clickHandler({ latlng: { lat: 40.7128, lng: -74.006 } });
    });

    expect(mockOnLocationSelect).not.toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Map onLocationSelect={mockOnLocationSelect} initialLocation={mockLocation} />
    );
    expect(container).toMatchSnapshot();
  });
});
