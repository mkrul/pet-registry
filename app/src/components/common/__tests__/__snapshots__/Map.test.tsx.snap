// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map > matches snapshot 1`] = `
<div>
  <div
    class="relative w-full h-[400px] rounded-lg overflow-hidden"
  >
    <div
      data-center="[40.7128,-74.006]"
      data-testid="map-container"
      data-zoom="15"
    >
      <div
        data-attribution="Â© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors"
        data-testid="tile-layer"
        data-url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      <div
        data-position="[40.7128,-74.006]"
        data-testid="map-marker"
      />
    </div>
    <div
      class="absolute inset-0 bg-white/75 z-[1000] flex items-center justify-center"
    >
      <div
        class="text-gray-300"
        data-size="32"
        data-testid="spinner"
      />
    </div>
  </div>
</div>
`;
